backends/optimized/ceed-opt-operator.c

```diff
@@ -18,12 +18,451 @@
+                                        lmode, op->inputfields[i].vec, opopt->evecs[iein],
+                                        request); CeedChk(ierr);
+        // Get evec
+        ierr = CeedVectorGetArrayRead(opopt->evecs[iein], CEED_MEM_HOST,
+                                      (const CeedScalar **) &opopt->edata[i]); CeedChk(ierr);
+        iein++;
+      } else {
+        // Active
+        // Restrict
+        ierr = CeedElemRestrictionApply(op->inputfields[i].Erestrict, CEED_NOTRANSPOSE,
+                                        lmode, invec, opopt->evecs[iein], request); CeedChk(ierr);
+        // Get evec
+        ierr = CeedVectorGetArrayRead(opopt->evecs[iein], CEED_MEM_HOST,
+                                      (const CeedScalar **) &opopt->edata[i]); CeedChk(ierr);
+        iein++;
+      }
+    } else {
+      // No restriction
+      CeedEvalMode emode = qf->inputfields[i].emode;
+      if (emode & CEED_EVAL_WEIGHT) {
+      } else {
+        // Passive
+        if (op->inputfields[i].vec) {
+          ierr = CeedVectorGetArrayRead(op->inputfields[i].vec, CEED_MEM_HOST,
+                                        (const CeedScalar **) &opopt->edata[i]); CeedChk(ierr);
+          // Active
+        } else {
+          ierr = CeedVectorGetArrayRead(invec, CEED_MEM_HOST,
+                                        (const CeedScalar **) &opopt->edata[i]); CeedChk(ierr);
+        }
+      }
+    }
+  }
+
+  // Output Evecs
+  for (CeedInt i=0,ieout=opopt->numein; i<qf->numoutputfields; i++) {
+    // Restriction
+    if (op->outputfields[i].Erestrict) {
+      ierr = CeedVectorGetArray(opopt->evecs[ieout], CEED_MEM_HOST,
+                                &opopt->edata[i + qf->numinputfields]); CeedChk(ierr);
+      ieout++;
+    } else {
+      // No restriction
+      // Passive
+      if (op->outputfields[i].vec) {
+        ierr = CeedVectorGetArray(op->outputfields[i].vec, CEED_MEM_HOST,
+                                  &opopt->edata[i + qf->numinputfields]); CeedChk(ierr);
+      } else {
+        // Active
+        ierr = CeedVectorGetArray(outvec, CEED_MEM_HOST,
+                                  &opopt->edata[i + qf->numinputfields]); CeedChk(ierr);
+      }
+    }
+  }
+
+  // Loop through elements
+  for (CeedInt b=0; b<nbatches; b++) {
+    if (b == nbatches - 1) batchsize = op->numelements - b*nelem;
+    e = b*nelem;
+    // Input basis apply if needed
+    for (CeedInt i=0; i<qf->numinputfields; i++) {
+      // Get elemsize
+      if (op->inputfields[i].Erestrict) {
+        elemsize = op->inputfields[i].Erestrict->elemsize;
+      } else {
+        elemsize = Q;
+      }
+      // Get emode, ncomp
+      CeedEvalMode emode = qf->inputfields[i].emode;
+      CeedInt ncomp = qf->inputfields[i].ncomp;
+      // Basis action
+      switch(emode) {
+      case CEED_EVAL_NONE:
+        break; // No action
+      case CEED_EVAL_INTERP:
+        data = op->inputfields[i].basis->data;
+        data->batchsize = batchsize;
```

Comment:
I'm really nervous about setting this here and then leaving it. Truthfully, I would rather _pad_ the input (by replicating data from the last element) so that the batch size is never mutated. Alternatively, we could add an explicit argument to `CeedBasisApply` for the batch size, but that would require adopting a convention for how to index across multiple elements (as a first or last index to the tensor).

Reply:
I found the output to be more of a problem because the tensor contract would try to write into unallocated memory if we aren't careful. Though, input can also be a problem for us because we might have an input with no restriction, so we don't have a natural way to pad that input without doing a special copy operation just for the last batch of elements. I added `batchsize` argument to try to prevent both these overwrites and special copies.

On line 262 I reset `batchsize` back to 1 on the last batch of elements. I am not fully happy with how I did it, but this field was my attempt to pass an argument to the `opt` backend `BasisApply` without reworking the arguments to the function for all backends.

Reply:
I see two viable approaches.
1. `batchsize` is an immutable property of a `CeedBasis` (though setting can be deferred). The caller can query it and is responsible for padding as needed (so that all lanes are valid).
2. `batchsize` is an argument of `CeedBasisApply` and existing callers will call with `batchsize=1`. (Variant: create new `CeedBasisApplyBatch`; `CeedBasisApply` can be implemented by calling `CeedBasisApplyBatch(batchsize=1, ...)`.)

In the former, the `CeedBasis` can statically optimize for a fixed batch size. In the latter, it will pick a path each time it is called. Since GPU batches might be very large [*], the second alternative may be preferable.

[*] The warp size is likely all that is needed in the "inner" dimension for a GPU, but each kernel launch will want to process many elements (likely the entire mesh). That said, an optimized GPU implementation will likely want to fuse BasisApply with other operations, in which case they would have a custom fused operation instead of calling via a generic `CeedBasisApplyBatch`.